### PR TITLE
Move EventProvider native layout to be driven by CMake configure

### LINKED ISSUE #2486: Move to cmake based incremental build from python
https://github.com/dotnet/coreclr/pull/2464 is unconditionally generating the ETW logging header files before each invocation. It completely kills incremental build - everything is rebuilt every time.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Move the dynamic layout of the EventProvider infastructure (excluding the C# EventSource file generation) to be driven by CMake instead of by custom scripting in our build.cmd and build.sh scripts. 

Fixes #2486

### CODE REVIEW

**Comment by sywhang** [comment]
cc @jorive

**Comment by jorive** [comment]
Have we verified that the generated files, before and after these changes, are the same?

**Comment by jkoritzinsky** [comment]
I'm verifying now.

**Comment by jkoritzinsky** [comment]
- EventPipe on Windows is identical
- Generated includes folder is identical
- Etw provider is identical
- PAL tests are identical
- Lttng provider: Diffs are only in the tracepoint headers. Look to be only ordering differences (contents are the same). I'll investigate to see if I can find the cause of the diff. Everything else is identical.
- Dummy provider: is identical

**Comment by jkoritzinsky** [comment]
I've verified that the Lttng provider files are the same other than the ordering differences.

**Comment by jkoritzinsky** [comment]
Looks like the CMake build was picking up a different version of Python on my test machine than my build on the branch. (Master was picking up 3.X by default, CMake was picking up 2.X). I think the ordering differences were caused by that.

File: src/scripts/genEventPipe.py [hunk 1]
```diff
@@ -274,7 +274,9 @@ def generateEventPipeCmakeFile(etwmanifest, eventpipe_directory, extern):
 """)
 
 def generateEventPipeHelperFile(etwmanifest, eventpipe_directory, extern):
-    with open_for_update(os.path.join(eventpipe_directory, "eventpipehelpers.cpp")) as helper:
+    eventpipehelpersPath = os.path.join(eventpipe_directory, "eventpipehelpers.cpp")
+    print(eventpipehelpersPath)
```
Comment (sywhang) [inline src/scripts/genEventPipe.py]: Did you mean to leave these prints here?

Comment (jkoritzinsky) [reply src/scripts/genEventPipe.py]: Yes I did. The way that the python script communicates to CMake which files it generates are by writing the path of the file to stdout, which CMake reads in to get the file paths.

**Comment by jkoritzinsky** [comment]
Test failures look to be networking infra failures. I'll retry the tests.

**Comment by jkoritzinsky** [comment]
Looks like the product and test build got split around #24452, so that's causing the test build failures.

**Comment by janvorli** [comment]
The dependency handling of ClrEtwAll.man seems to be broken. When I build it on my Ubuntu 16.04, then modify the src/vm/ClrEtwAll.man by adding a new event and build again, I can see that the clralltestevents.cpp was regenerated, but it was not recompiled. In fact, nothing was recompiled, which doesn't seem right.
I was thinking that the clralltestevents.cpp issue may be related to a problem that exists with add_custom_command when it generates header files or other files that are not directly processed by compilers and that we need to mark the file explicitly as generated and add a custom _target that the target using the generated source would depend on. However, I was wrong and it doesn't fix the issue.

**Comment by jkoritzinsky** [comment]
Yeah at the moment it should regenerate on reconfigure. I was trying to avoid having to specify all of the header dependencies to CMake but it looks like I might need to do that.

**Comment by janvorli** [comment]
I believe that we need to make sure incremental build works reliably. This moves us to the other extreme than the preexisting state. Without your change it compiles these files on every build. With your change, we need to do a clean build after changes to src/vm/ClrEtwAll.man.

**Comment by jkoritzinsky** [comment]
I'm working on getting a good incremental build story working for it now.

**Comment by jkoritzinsky** [comment]
@janvorli I got incremental builds working (at least locally) if you want to take a look when you have a chance.

**Comment by janvorli** [comment]
@jkoritzinsky it still doesn't work. While it correctly regenerated all the files and headers after I've modified the ClrEtwAll.man, it didn't recompile anything.

**Comment by janvorli** [comment]
One thing that is missing here is marking of the generated headers using `set_source_files_properties(your_header_file_path PROPERTIES GENERATED TRUE)`. While this should not be needed for cpp files, it is needed for .h ones.

File: src/vm/ClrEtwAll.man [hunk 2]
```diff
@@ -2398,10 +2398,10 @@
                            task="GarbageCollection"
                            symbol="GCStart" message="$(string.RuntimePublisher.GCStartEventMessage)"/>
 
-                    <event value="1" version="1" level="win:Informational"  template="GCStart_V1"
+                    <!-- <event value="1" version="1" level="win:Informational"  template="GCStart_V1"
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/vm/ClrEtwAll.man]: Yep. Looks like I missed that one.

**Comment by jkoritzinsky** [comment]
I've marked the headers as generated and it looks like it works for me (adding and removing events cause recompiles of the product binaries and the PAL test).

**Comment by jkoritzinsky** [comment]
All failures are timeouts in Helix tracked by https://github.com/dotnet/core-eng/issues/6295, so I'm going to merge.

**Pull request merged**